clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -128,23 +143,75 @@ record_filter_t::parallel_shard_error(void *shard_data)
     return per_shard->error;
 }
 
+bool
+record_filter_t::write_trace_entry(per_shard_t *shard, const trace_entry_t &entry)
+{
+    if (!shard->writer->write((char *)&entry, sizeof(entry))) {
+        shard->error = "Failed to write to output file " + shard->output_path;
+        success_ = false;
+        return false;
+    }
+    ++shard->output_entry_count;
+    return true;
+}
+
 bool
 record_filter_t::parallel_shard_memref(void *shard_data, const trace_entry_t &entry)
 {
     per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
+    ++per_shard->input_entry_count;
     bool output = true;
-    for (int i = 0; i < (int)filters_.size(); ++i) {
-        if (!filters_[i]->parallel_shard_filter(entry, filter_shard_data_[i]))
+    if (stop_timestamp_us_ == 0 ||
+        per_shard->shard_stream->get_last_timestamp() < stop_timestamp_us_) {
+        for (int i = 0; i < (int)filters_.size(); ++i) {
+            if (!filters_[i]->parallel_shard_filter(entry,
+                                                    per_shard->filter_shard_data[i]))
+                output = false;
+        }
+    } else {
+        if (per_shard->enabled) {
+            per_shard->enabled = false;
+            trace_entry_t filter_boundary_entry = { TRACE_TYPE_MARKER,
+                                                    TRACE_MARKER_TYPE_FILTER_BOUNDARY,
+                                                    { 0 } };
+            if (!write_trace_entry(per_shard, filter_boundary_entry))
+                return false;
+        }
+    }
+
+    // Optimize space by outputting the unit header only if we are outputting something
+    // from that unit.
+    if (entry.type == TRACE_TYPE_MARKER) {
+        switch (entry.size) {
+        case TRACE_MARKER_TYPE_TIMESTAMP:
+            // No need to remember the previous unit's header anymore. We're in the
+            // next unit now.
+            // XXX: it may happen that we never output a unit header due to this
+            // optimization. We should ensure that we output it atleast once. We
```

Comment:
What is the reason to output "at least once"?

Reply:
Because if there's no timestamp entry at all, some of our assumptions fail about the trace format (e.g. "everything up to the timestamp entry is the shard header").

This is still in line with what we discussed: if a record_filter_func_t wants to filter out timestamp entries, it can still do it. This "at least once" requirement is for the space optimization here.